leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/ClientStatusTracker.java

```diff
@@ -0,0 +1,114 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Bosch Software Innovations GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ * Contributors:
+ *     Bosch Software Innovations GmbH - initial API
+ *******************************************************************************/
+package org.eclipse.leshan.server.queue;
+
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.eclipse.leshan.server.registration.Registration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Tracks the status of each LWM2M client registered with Queue mode binding. Also ensures that the
+ * {@link QueueModeClientStatusListener} are notified on state changes only for those LWM2M clients registered using
+ * Queue mode binding.
+ *
+ * @see QueueModeClientState
+ */
+public final class ClientStatusTracker {
+    private static final Logger LOG = LoggerFactory.getLogger(ClientStatusTracker.class);
+    private static ClientStatusTracker instance = null;
+    private final ConcurrentMap<String, QueueModeClientState> clientStatus = new ConcurrentHashMap<>();
+    private final List<QueueModeClientStatusListener> listeners = new CopyOnWriteArrayList<>();
+
+    private ClientStatusTracker() {
+        // Exists to prevent instantiation.
+    }
+
+    public static synchronized ClientStatusTracker getInstance() {

```

Comment:
The reason why it is made singleton is that, no matter where this tracker is used (like `LeshanServer`, `ObservationService`, `RegistrationService `etc.) they all add `PresenceListener` to the same list. If not, we have to make sure that the tracker is instantiated only once manually.

Any problem with singletons?